### PR TITLE
GH-1692: Add some safety to ASN's binary_to_int64 to avoid bad shifts

### LINKED ISSUE #1692: Crash 4.1.0 - snmp_pac.cc:849:16: runtime error: shift exponent 640 is too large for 64-bit type 'long unsigned int'
This one seems to be a new - somehow I don't have a coredump of this in spool/tmp - I am pasting this stack trace out of worker-1-1/stderr.log

```
/usr/local/src/zeek-4.1.0-rc1/build/src/analyzer/protocol/snmp/snmp_pac.cc:849:16: runtime error: shift exponent 640 is too large for 64-bit type 'long unsigned int'
    #0 0x55d9b8db3870 in binpac::SNMP::binary_to_int64(binpac::datastring<unsigned char> const&) /usr/local/src/zeek-4.1.0-rc1/build/src/analyzer/protocol/snmp/snmp_pac.cc:849
    #1 0x55d9b8db01dc in binpac::SNMP::ASN1EncodingMeta::Parse(unsigned char const*, unsigned char const*) /usr/local/src/zeek-4.1.0-rc1/build/src/analyzer/protocol/snmp/snmp_pac.cc:438
    #2 0x55d9b8db0bb1 in binpac::SNMP::ASN1SequenceMeta::Parse(unsigned char const*, unsigned char const*, binpac::SNMP::ContextSNMP*) /usr/local/src/zeek-4.1.0-rc1/build/src/analyzer/protocol/snmp/snmp_pac.cc:522
    #3 0x55d9b8db4da0 in binpac::SNMP::TopLevelMessage::Parse(unsigned char const*, unsigned char const*, binpac::SNMP::ContextSNMP*) /usr/local/src/zeek-4.1.0-rc1/build/src/analyzer/protocol/snmp/snmp_pac.cc:1001
    #4 0x55d9b8dc17c3 in binpac::SNMP::SNMP_Flow::NewData(unsigned char const*, unsigned char const*) /usr/local/src/zeek-4.1.0-rc1/build/src/analyzer/protocol/snmp/snmp_pac.cc:2543
    #5 0x55d9b8dabf62 in binpac::SNMP::SNMP_Conn::NewData(bool, unsigned char const*, unsigned char const*) /usr/local/src/zeek-4.1.0-rc1/build/src/analyzer/protocol/snmp/snmp_pac.cc:46
    #6 0x55d9b8da1ebb in zeek::analyzer::snmp::SNMP_Analyzer::DeliverPacket(int, unsigned char const*, bool, unsigned long, zeek::IP_Hdr const*, int) /usr/local/src/zeek-4.1.0-rc1/src/analyzer/protocol/snmp/SNMP.cc:36
    #7 0x55d9b87b8df2 in zeek::analyzer::Analyzer::NextPacket(int, unsigned char const*, bool, unsigned long, zeek::IP_Hdr const*, int) /usr/local/src/zeek-4.1.0-rc1/src/analyzer/Analyzer.cc:226
    #8 0x55d9b87b9757 in zeek::analyzer::Analyzer::ForwardPacket(int, unsigned char const*, bool, unsigned long, zeek::IP_Hdr const*, int) /usr/local/src/zeek-4.1.0-rc1/src/analyzer/Analyzer.cc:310
    #9 0x55d9b8f10af7 in zeek::packet_analysis::UDP::UDPAnalyzer::DeliverPacket(zeek::Connection*, double, bool, int, zeek::Packet*) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/protocol/udp/UDP.cc:224
    #10 0x55d9b8f02793 in zeek::packet_analysis::IP::IPBasedAnalyzer::AnalyzePacket(unsigned long, unsigned char const*, zeek::Packet*) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/protocol/ip/IPBasedAnalyzer.cc:96
    #11 0x55d9b8798692 in zeek::packet_analysis::Analyzer::ForwardPacket(unsigned long, unsigned char const*, zeek::Packet*, unsigned int) const /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/Analyzer.cc:97
    #12 0x55d9b8f001f8 in zeek::packet_analysis::IP::IPAnalyzer::AnalyzePacket(unsigned long, unsigned char const*, zeek::Packet*) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/protocol/ip/IP.cc:260
    #13 0x55d9b8798692 in zeek::packet_analysis::Analyzer::ForwardPacket(unsigned long, unsigned char const*, zeek::Packet*, unsigned int) const /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/Analyzer.cc:97
    #14 0x55d9b87a2342 in zeek::packet_analysis::Manager::ProcessInnerPacket(zeek::Packet*) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/Manager.cc:122
    #15 0x55d9b8f4d4d2 in zeek::packet_analysis::IPTunnel::IPTunnelAnalyzer::ProcessEncapsulatedPacket(double, zeek::Packet const*, unsigned int, unsigned int, unsigned char const*, int, std::shared_ptr<zeek::EncapsulationStack>, zeek::EncapsulatingConn const&) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/protocol/iptunnel/IPTunnel.cc:172
    #16 0x55d9b8f4c62c in zeek::packet_analysis::IPTunnel::IPTunnelAnalyzer::AnalyzePacket(unsigned long, unsigned char const*, zeek::Packet*) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/protocol/iptunnel/IPTunnel.cc:87
    #17 0x55d9b8798869 in zeek::packet_analysis::Analyzer::ForwardPacket(unsigned long, unsigned char const*, zeek::Packet*) const /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/Analyzer.cc:103
    #18 0x55d9b8f49137 in zeek::packet_analysis::GRE::GREAnalyzer::AnalyzePacket(unsigned long, unsigned char const*, zeek::Packet*) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/protocol/gre/GRE.cc:217
    #19 0x55d9b8798692 in zeek::packet_analysis::Analyzer::ForwardPacket(unsigned long, unsigned char const*, zeek::Packet*, unsigned int) const /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/Analyzer.cc:97
    #20 0x55d9b8f001f8 in zeek::packet_analysis::IP::IPAnalyzer::AnalyzePacket(unsigned long, unsigned char const*, zeek::Packet*) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/protocol/ip/IP.cc:260
    #21 0x55d9b8798692 in zeek::packet_analysis::Analyzer::ForwardPacket(unsigned long, unsigned char const*, zeek::Packet*, unsigned int) const /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/Analyzer.cc:97
    #22 0x55d9b8ede717 in zeek::packet_analysis::Ethernet::EthernetAnalyzer::AnalyzePacket(unsigned long, unsigned char const*, zeek::Packet*) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/protocol/ethernet/Ethernet.cc:56
    #23 0x55d9b8798692 in zeek::packet_analysis::Analyzer::ForwardPacket(unsigned long, unsigned char const*, zeek::Packet*, unsigned int) const /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/Analyzer.cc:97
    #24 0x55d9b87a1fe4 in zeek::packet_analysis::Manager::ProcessPacket(zeek::Packet*) /usr/local/src/zeek-4.1.0-rc1/src/packet_analysis/Manager.cc:109
    #25 0x55d9b811b6b9 in zeek::run_state::detail::dispatch_packet(zeek::Packet*, zeek::iosource::PktSrc*) /usr/local/src/zeek-4.1.0-rc1/src/RunState.cc:257
    #26 0x55d9b860d63e in zeek::iosource::PktSrc::Process() /usr/local/src/zeek-4.1.0-rc1/src/iosource/PktSrc.cc:155
    #27 0x55d9b811be09 in zeek::run_state::detail::run_loop() /usr/local/src/zeek-4.1.0-rc1/src/RunState.cc:315
    #28 0x55d9b7a721a0 in main /usr/local/src/zeek-4.1.0-rc1/src/main.cc:59
    #29 0x7f4ada154bf6 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21bf6)
    #30 0x55d9b7a71ae9 in _start (/usr/local/zeek-4.1.0-rc1/bin/zeek+0x1182ae9)

```

### PR DESCRIPTION
**Pull request by timwoj**
Fixes #1692

### CODE REVIEW

**Comment by 0xxon** [comment]
The one slight concern I have about this one is what happens in cases where the number that we are converting is actually larger than 64 bits. ASN.1 does allow that - and while we might not have it in our test traffic, I can see that larger numbers might actually appear in legitimate traffic.

If that happens, that means that after this change the parsing of the traffic is just stopped, right? I kind of wonder if it might be better to throw a warning in these cases - and to return a truncated length. But - I am also not quite sure.

**Comment by timwoj** [comment]
> The one slight concern I have about this one is what happens in cases where the number that we are converting is actually larger than 64 bits. ASN.1 does allow that - and while we might not have it in our test traffic, I can see that larger numbers might actually appear in legitimate traffic.

Without this patch, we just hit some undefined behavior and shift more than we should.

> If that happens, that means that after this change the parsing of the traffic is just stopped, right?

Right. As far as I can tell from binpac, throwing an exception just unwinds the parsing and the analysis stops at the level that the exception was thrown.

> I kind of wonder if it might be better to throw a warning in these cases - and to return a truncated length. But - I am also not quite sure.

That sounds fine to me and was actually the first way that I implemented it, but I was worried about the number of warnings that would get generated. Maybe a weird, since those are rate limited?

File: src/analyzer/protocol/asn1/asn1.pac [hunk 1]
```diff
@@ -87,10 +87,18 @@ type Array = record {
 
 ############################## ASN.1 Conversion Functions
 
+# Converts an 8-byte string into an int64. If this string is longer than
+# 8 bytes, it throws an exception.
```
Comment (0xxon) [inline src/analyzer/protocol/asn1/asn1.pac:91]: This should be "a weird", right? :)

**Review by 0xxon — APPROVED** [review]

File: src/analyzer/protocol/asn1/asn1.pac (hunk 1 above)
Comment (timwoj) [reply src/analyzer/protocol/asn1/asn1.pac]: Fixed. I'm going to let CI chew on this and merge it if everything passes.

**Pull request merged**